### PR TITLE
Implement composite rules, joined via and()

### PR DESCRIPTION
**Pull request by bogsi17**

### CODE REVIEW

**Comment by codecholeric** [comment]
Thanks a lot :smiley: Looks generally good, I've added some minor comments. Also as you can see, the tests have failed. These are ArchUnit's own ArchUnit tests :wink: that want that you add `@PublicAPI` to public methods you want to expose (it's mainly a check to remind developers to really think carefully about exposed API that we have to support for a long time). So if it's `@PublicAPI(usage = ACCESS)`, public classes and methods should also be final (since the API is not intended for inheritance).

Also it would be great if you could check out https://github.com/TNG/ArchUnit/blob/master/CONTRIBUTING.md
In particular it's necessary that you sign off your commit (check 'sign-off' in IntelliJ or commit via `git commit -s` from the command line). Referencing the issue would be great, too!
Anyway, thanks a lot for contributing, I really appreciate it!! So I hope all those nagging little comments don't discourage you :wink:

**Comment by codecholeric** [comment]
PS: Rebasing would be awesome, too :wink: I generally like a linear history where possible...

**Comment by bogsi17** [comment]
Hi, thanks for the feedback! Where can I find your comments? Can't see them in the 'Files Changed' section.

And yeah, I really should have read this document before :D

**Comment by codecholeric** [comment]
Ah sorry, I forgot to click "submit review" :flushed:

File: archunit-example/example-plain/src/test/java/com/tngtech/archunit/exampletest/CompositeArchRulesTest.java [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+import org.junit.experimental.categories.Category;
+
+import static com.tngtech.archunit.library.GeneralCodingRules.*;
+
+@Category(Example.class)
+public class CompositeArchRulesTest {
+    private final JavaClasses classes = new ClassFileImporter().importPackages("com.tngtech.archunit.example");
+
+    @Test
+    public void no_classes_should_access_standard_streams_and_throw_generic_exceptions() {
```
[the remark this answers is not in the document]
Comment (bogsi17) [reply archunit-example/example-plain/src/test/java/com/tngtech/archunit/exampletest/CompositeArchRulesTest.java]: yeah, changed it to `no_classes_should_A_or_B`

File: archunit-integration-test/src/test/java/com/tngtech/archunit/integration/ExamplesIntegrationTest.java [hunk 2]
```diff
@@ -184,6 +185,41 @@
                 .toDynamicTests();
     }
 
+    @TestFactory
+    Stream<DynamicTest> CompositeArchRulesTest() {
+        return ExpectedTestFailures
+            .forTests(CompositeArchRulesTest.class)
+
+            .ofRule("no classes should access standard streams and no classes should throw generic exceptions")
+            .by(callFromMethod(ClassViolatingCodingRules.class, "printToStandardStream")
```
[the remark this answers is not in the document]
Comment (bogsi17) [reply archunit-integration-test/src/test/java/com/tngtech/archunit/integration/ExamplesIntegrationTest.java]: ok

File: archunit/src/main/java/com/tngtech/archunit/lang/CompositeArchRule.java [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+import com.google.common.base.Joiner;
+import com.google.common.collect.Iterables;
+import com.tngtech.archunit.core.domain.JavaClasses;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+public class CompositeArchRule implements ArchRule {
+    private List<ArchRule> rules;
```
[the remark this answers is not in the document]
Comment (bogsi17) [reply archunit/src/main/java/com/tngtech/archunit/lang/CompositeArchRule.java]: ok

**Comment by codecholeric** [comment]
I've added support for priorities to `CompositeArchRule`, other than that, it looks good now :smile: 
Gonna merge it...

**Pull request merged**